weblate/trans/views/helper.py

```diff
@@ -0,0 +1,235 @@
+            raise Exception('Invalid language!')
+    except Exception:
+        # Ignore failure on activating language
+        django.utils.translation.activate('en')
+
+
+def import_message(request, count, message_none, message_ok):
+    if count == 0:
+        messages.warning(request, message_none)
+    else:
+        messages.success(request, message_ok % count)
+
+def get_formatted_translation_file(translation, fmt, form_cleaned_data, http_response=False, tmpdir=None):
+    try:
+        exporter = get_exporter(fmt)(translation=translation)
+    except KeyError:
+        raise Http404('File format not supported')
+
+    units = translation.unit_set.search(
+        form_cleaned_data,
+        translation=translation,
+    )
+
+    if units is None:
+        units = translation.unit_set.all()
+    
+    exporter.add_units(units)
+
+    filename_template = '{{project}}-{0}-{{language}}.{{extension}}'.format(
+        translation.component.slug
+    )
+    
+    if http_response:
+        return exporter.get_response(filename_template)
+    else:    
+        filename = exporter.getFileName(filename_template)
+        filePath = tmpdir + '/' + filename
+        tmpFilePath = exporter.writeToFile(filePath)
+
+        return dict([('srcFilePath', filePath), ('destFileName', filename)])
+    
+def get_translation_file(translation, fmt, form_cleaned_data, tmpdir):
+    if fmt is not None:
+        return get_formatted_translation_file(translation, fmt, form_cleaned_data, tmpdir=tmpdir)
+
+    # Force flushing pending units
+    translation.commit_pending('download', None)
+
+    srcfilename = translation.get_filename()
+
+    return srcfilename
+
+def download_translation_file(translation, fmt=None, form_cleaned_data=None):
+    if fmt is not None:
+        return get_formatted_translation_file(translation, fmt, form_cleaned_data, http_response=True)
+
+
+    # Force flushing pending units
+    translation.commit_pending('download', None)
+
+    srcfilename = translation.get_filename()
+
+    # Construct file name (do not use real filename as it is usually not
+    # that useful)
+
+    filename = translation_filename(translation)
+
+    # Create response
+    with open(srcfilename) as handle:
+        response = HttpResponse(
+            handle.read(),
+            content_type=translation.store.mimetype
+        )
+
+    # Fill in response headers
+    response['Content-Disposition'] = 'attachment; filename={0}'.format(
+        filename
+    )
+
+    return response
+
+def translation_filename(translation):
+    return '{0}-{1}-{2}.{3}'.format(
+        translation.component.project.slug,
+        translation.component.slug,
+        translation.language.code,
+        translation.store.extension
+    )
+
+def download_translation_files(translations, fmt=None, form_cleaned_data=None):
+    dirname = 0
+    try:
+        dirname = tempfile.mkdtemp()
+        s = BytesIO()
+
+        with zipfile.ZipFile(s, 'w') as translationsZip:
```

Comment:
Better to construct HttpResponse here and use it as handle for ZipFile. That avoids need to copy zip file content to response.